### PR TITLE
added first version of TPC QC cluster workflow

### PR DESCRIPTION
**Pull request by tklemenz**
Added a first version of a QC workflow for TPC clusters.

### CODE REVIEW

**Comment by tklemenz** [comment]
This pull request can only be merged after https://github.com/AliceO2Group/AliceO2/pull/4006.

File: Modules/TPC/src/Clusters.cxx [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+  delete (qMaxAHisto);
+  delete (qMaxCHisto);
+  delete (qTotAHisto);
+  delete (qTotCHisto);
+  delete (sigmaTimeAHisto);
+  delete (sigmaTimeCHisto);
+  delete (sigmaPadAHisto);
+  delete (sigmaPadCHisto);
+  delete (timeBinAHisto);
+  delete (timeBinCHisto);
```
Comment (wiechula) [inline Modules/TPC/src/Clusters.cxx:252]: With [This PR](https://github.com/AliceO2Group/AliceO2/pull/4070) you can do:

```suggestion
  //TODO: Method create-copy-delete needs to be improved
  int histIter = 0;
  for (const auto& calPadWapper : mWrapperVector) {
    auto& histA =  *mHistoVector[histIter].get();
    auto& histC =  *mHistoVector[histIter + 1].get();
    const auto& calPad = calPadWrapper->getObj();

    o2::tpc::painter::fillHistogram2D(histA, calPad, ASide);
    o2::tpc::painter::fillHistogram2D(histC, calPad, CSide);
   
     histIter += 2;
  }

```

**Comment by tklemenz** [comment]
Hi @wiechula, I needed to make a few minor changes in your suggestions for the Cluster.cxx to make it work but it is running and publishing now as intended.

**Comment by tklemenz** [comment]
The checks fail at the moment because https://github.com/AliceO2Group/AliceO2/pull/4006 needs to be merged first.

**Comment by stheckel** [comment]
Hi,
the O2 PR has been merged. Can we trigger the re-testing of the checks somehow? Then this PR could be merged as well.
Cheers,
Stefan

**Comment by tklemenz** [comment]
@wiechula, do you know why mac does not like the use of the fillHistogram2D function of the painter in the monitorData function?

File: Modules/TPC/src/Clusters.cxx [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+#include "TPCBase/Painter.h"
+#include "DataFormatsTPC/ClusterNative.h"
+#include "DataFormatsTPC/ClusterNativeHelper.h"
+#include "SimulationDataFormat/MCTruthContainer.h"
+#include "DataFormatsTPC/Constants.h"
+
+// QC includes
+#include "QualityControl/QcInfoLogger.h"
+#include "TPC/Clusters.h"
+
```
Comment (wiechula) [inline Modules/TPC/src/Clusters.cxx:49]: can you try to add
`template void o2::tpc::painter::fillHistogram2D<float>(TH2& h2D, const CalDet<float>& calDet, Side side);`
to see if this solved the mac issue?

Comment (wiechula) [reply Modules/TPC/src/Clusters.cxx:49]: If so I will add this in o2 and then it can be removed again, but it will be easier to quickly test it here.

Comment (tklemenz) [reply Modules/TPC/src/Clusters.cxx:49]: Hi @wiechula, I inserted the line in Cluster.cxx but get an error:
`Scanning dependencies of target QcTPC
[ 95%] Building CXX object Modules/TPC/CMakeFiles/QcTPC.dir/src/Clusters.cxx.o
[ 96%] Building CXX object Modules/TPC/CMakeFiles/QcTPC.dir/G__QcTPC.cxx.o
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
 template void o2::tpc::painter::fillHistogram2D<float>(TH2& h2D, const CalDet<float>& calDet, Side side);
                                                                                                        ^
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx: In instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]':
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104:   required from here
/home/tklemenz/AliSoftware/sw/SOURCES/QualityControl/clusterTask/0/Modules/TPC/src/Clusters.cxx:50:104: error: explicit instantiation of 'void o2::tpc::painter::fillHistogram2D(TH2&, const o2::tpc::CalDet<T>&, o2::tpc::Side) [with T = float]' but no definition available [-fpermissive]
Modules/TPC/CMakeFiles/QcTPC.dir/build.make:147: recipe for target 'Modules/TPC/CMakeFiles/QcTPC.dir/src/Clusters.cxx.o' failed
make[2]: *** [Modules/TPC/CMakeFiles/QcTPC.dir/src/Clusters.cxx.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:2143: recipe for target 'Modules/TPC/CMakeFiles/QcTPC.dir/all' failed
make[1]: *** [Modules/TPC/CMakeFiles/QcTPC.dir/all] Error 2
Makefile:157: recipe for target 'all' failed
make: *** [all] Error 2`

Comment (tklemenz) [reply Modules/TPC/src/Clusters.cxx:49]: I put the line in but get an error, see [here](https://cernbox.cern.ch/index.php/s/6eR2RgDvaU9nNFx).

Comment (wiechula) [reply Modules/TPC/src/Clusters.cxx:49]: My bad. This cannot work, indeed. Then let's wait for https://github.com/AliceO2Group/AliceO2/pull/4111 and see if this solves the issue.

**Comment by tklemenz** [comment]
The checks have passed, this can be merged.

File: Modules/TPC/run/tpcQCClusters_direct.json [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        "cycleDurationSeconds": "10",
+        "maxNumberCycles": "-1",
+        "dataSource": {
+          "type": "direct",
+          "query" : "input0:TPC/CLUSTERNATIVE/0;input1:TPC/CLUSTERNATIVE/1;input2:TPC/CLUSTERNATIVE/2;input3:TPC/CLUSTERNATIVE/3;input4:TPC/CLUSTERNATIVE/4;input5:TPC/CLUSTERNATIVE/5;input6:TPC/CLUSTERNATIVE/6;input7:TPC/CLUSTERNATIVE/7;input8:TPC/CLUSTERNATIVE/8;input9:TPC/CLUSTERNATIVE/9;input10:TPC/CLUSTERNATIVE/10;input11:TPC/CLUSTERNATIVE/11;input12:TPC/CLUSTERNATIVE/12;input13:TPC/CLUSTERNATIVE/13;input14:TPC/CLUSTERNATIVE/14;input15:TPC/CLUSTERNATIVE/15;input16:TPC/CLUSTERNATIVE/16;input17:TPC/CLUSTERNATIVE/17;input18:TPC/CLUSTERNATIVE/18;input19:TPC/CLUSTERNATIVE/19;input20:TPC/CLUSTERNATIVE/20;input21:TPC/CLUSTERNATIVE/21;input22:TPC/CLUSTERNATIVE/22;input23:TPC/CLUSTERNATIVE/23;input24:TPC/CLUSTERNATIVE/24;input25:TPC/CLUSTERNATIVE/25;input26:TPC/CLUSTERNATIVE/26;input27:TPC/CLUSTERNATIVE/27;input28:TPC/CLUSTERNATIVE/28;input29:TPC/CLUSTERNATIVE/29;input30:TPC/CLUSTERNATIVE/30;input31:TPC/CLUSTERNATIVE/31;input32:TPC/CLUSTERNATIVE/32;input33:TPC/CLUSTERNATIVE/33;input34:TPC/CLUSTERNATIVE/34;input35:TPC/CLUSTERNATIVE/35"
```
Comment (knopers8) [inline Modules/TPC/run/tpcQCClusters_direct.json:35]: If you don't need all of those inputs at the same time, but they can arrive one after another, you can just use `"input:TPC/CLUSTERNATIVE"` - this should request this data with any subspec. That might also decrease memory and CPU usage coming from DPL trying to organise all of those inputs.

**Pull request merged**